pkg/solana/chainwriter/chain_writer.go

```diff
@@ -220,6 +225,76 @@ func (s *SolanaChainWriterService) FilterLookupTableAddresses(
 	return filteredLookupTables
 }
 
+// CreateATAs first checks if a specified location exists, then checks if the accounts derived from the
+// ATALookups in the ChainWriter's configuration exist on-chain and creates them if they do not.
+func CreateATAs(ctx context.Context, args any, lookups []ATALookup, derivedTableMap map[string]map[string][]*solana.AccountMeta, client client.MultiClient, idl string, feePayer solana.PublicKey) ([]solana.Instruction, error) {
+	createATAInstructions := []solana.Instruction{}
+	for _, lookup := range lookups {
+		// Check if location exists
+		if lookup.Location != "" {
+			// TODO refactor GetValuesAtLocation to not return an error if the field doesn't exist
+			_, err := GetValuesAtLocation(args, lookup.Location)
+			if err != nil {
+				// field doesn't exist, so ignore ATA creation
+				if errors.Is(err, errFieldNotFound) {
+					continue
+				}
+				return nil, fmt.Errorf("error getting values at location: %w", err)
+			}
+		}
+		walletAddresses, err := GetAddresses(ctx, args, []Lookup{lookup.WalletAddress}, derivedTableMap, client)
+		if err != nil {
+			return nil, fmt.Errorf("error resolving wallet address: %w", err)
+		}
+		if len(walletAddresses) != 1 {
+			return nil, fmt.Errorf("expected exactly one wallet address, got %d", len(walletAddresses))
+		}
+		wallet := walletAddresses[0].PublicKey
+
+		tokenPrograms, err := GetAddresses(ctx, args, []Lookup{lookup.TokenProgram}, derivedTableMap, client)
```

Comment:
In CCIP, we know the location of tokenPrograms will be an array.
I recall, you already added the support to flatten the array into multiple values right?
Note that in CCIP case, the leaf node isn't the array field, but an intermediate path is.

Any2SVMRampMessage.TokenAmounts.DestTokenAddress
Here, the TokenAmounts is the array field.

I wasn't sure if current code will enumerate all the TokenAmounts here.

Reply:
Yes I added in this functionality it will work as expected